Firestore/core/test/firebase/firestore/util/async_queue_test.cc

```diff
@@ -0,0 +1,248 @@
+// associated future (with timeout).
+bool WaitForTestToFinish(SignalT* const signal) {
+  return signal->get_future().wait_for(kTimeout) ==
+         std::future_status::ready;
+}
+
+}  // namespace
+
+TEST(AsyncQueue, Enqueue) {
+  auto signal_finished = CreateSignal();
+
+  auto queue = Queue();
+  queue.Enqueue([&] { signal_finished(); });
+
+  EXPECT_TRUE(WaitForTestToFinish(&signal_finished));
+}
+
+TEST(AsyncQueue, EnqueueDisallowsEnqueuedTasksToUseEnqueue) {
+  auto signal_finished = CreateSignal();
+
+  auto queue = Queue();
+  queue.Enqueue([&] {  // clang-format off
+    EXPECT_ANY_THROW(queue.Enqueue([&] { signal_finished(); }););
+    // clang-format on
+  });
+
+  WaitForTestToFinish(&signal_finished);
+}
+
+TEST(AsyncQueue, EnqueueAllowsEnqueuedTasksToUseEnqueueUsingSameQueue) {
+  auto signal_finished = CreateSignal();
+
+  auto queue = Queue();
+  queue.Enqueue([&] {  // clang-format off
+    queue.EnqueueAllowingSameQueue([&] { signal_finished(); });
+    // clang-format on
+  });
+
+  EXPECT_TRUE(WaitForTestToFinish(&signal_finished));
+}
+
+TEST(AsyncQueue, SameQueueIsAllowedForUnownedActions) {
+  auto signal_finished = CreateSignal();
+  auto queue = Queue();
+
+  using WrapT = std::pair<AsyncQueue*, std::packaged_task<void()>*>;
+  WrapT wrap{&queue, &signal_finished};
+  dispatch_async_f(underlying_queue, &wrap, [](void* const raw_wrap) {
+    auto unwrap = static_cast<const WrapT*>(raw_wrap);
+    unwrap->first->Enqueue([unwrap] { (*unwrap->second)(); });
+  });
+
+  EXPECT_TRUE(WaitForTestToFinish(&signal_finished));
+}
+
+TEST(AsyncQueue, EnqueueSync) {
+  bool finished = false;
+
+  auto queue = Queue();
+  queue.EnqueueSync([&] { finished = true; });
+
+  EXPECT_TRUE(finished);
+}
+
+TEST(AsyncQueue, EnqueueSyncDisallowsEnqueuedTasksToUseEnqueue) {
+  auto queue = Queue();
+  queue.EnqueueSync([&] {  // clang-format off
+    EXPECT_ANY_THROW(queue.EnqueueSync([] {}););
+    // clang-format on
+  });
+}
+
+TEST(AsyncQueue, EnterCheckedOperationDisallowsNesting) {
+  auto queue = Queue();
+  queue.EnqueueSync([&] {
+    EXPECT_ANY_THROW(queue.EnterCheckedOperation([&] {}););
+  });
+}
+
+TEST(AsyncQueue, VerifyIsCurrentQueueRequiresCurrentQueue) {
+  ASSERT_NE(underlying_queue, dispatch_get_main_queue());
+  EXPECT_ANY_THROW(Queue().VerifyIsCurrentQueue());
+}
+
+TEST(AsyncQueue, VerifyIsCurrentQueueRequiresOperationInProgress) {
+  auto queue = Queue();
+  dispatch_sync_f(underlying_queue, &queue, [](void* const raw_queue) {
+    EXPECT_ANY_THROW(
+        static_cast<AsyncQueue*>(raw_queue)->VerifyIsCurrentQueue());
+  });
+}
+
+TEST(AsyncQueue, VerifyIsCurrentQueueWorksWithOperationInProgress) {
+  auto queue = Queue();
+  queue.EnqueueSync([&] { EXPECT_NO_THROW(queue.VerifyIsCurrentQueue()); });
+}
+
+TEST(AsyncQueue, CanScheduleOperationsInTheFuture) {
+  auto signal_finished = CreateSignal();
+  std::string steps;
+
+  auto queue = Queue();
+  queue.Enqueue([&steps] { steps += '1'; });
```

Comment:
My reading of the [docs](https://developer.apple.com/legacy/library/documentation/Darwin/Reference/ManPages/man3/dispatch_async_f.3.html) is that accessing `steps` from several enqueued tasks is thread-safe:

> The dispatch_async() function may be used to implement deferred critical sections [...] queues support both asynchronous and synchronous operations, and those operations are ordered by definition.

But I'd appreciate if @wilhuff could confirm once he gets back from vacation.